### PR TITLE
My code implemented for the LSTM model

### PR DESCRIPTION
**Pull request by mingdinh**
- The cleaning file I forgot to delete the UTF-8 BOM in the dataset (\xef \xbf \xbd ...) so I add it to the clean_dataset file.
- 0.2% of tweets are empty after cleaning. So I delete the empty tweet in the preload.py. In this file, I also take 10% of the data set for training (it already took me over an hour to scan with 10% of the data set).

I use Glove vector for embedding.

I used an LSTM model.

### CODE REVIEW

File: clean_dataset.py [hunk 1]
```diff
@@ -1,5 +1,3 @@
-"""Script to clean the dataset."""
-
```
Comment (ElieKadoche) [inline clean_dataset.py]: Why do you remove the docstring?

File: requirements.txt [hunk 2]
```diff
@@ -1,6 +1,4 @@
-bs4
-nltk
```
Comment (ElieKadoche) [inline requirements.txt:2]: Why do you remove these?

File: clean_dataset.py [hunk 3]
```diff
@@ -9,14 +7,16 @@
...
 
+from bs4 import BeautifulSoup
+from nltk.tokenize import WordPunctTokenizer
+from sklearn.model_selection import train_test_split
 def cleaner(data):
     """Clean data.
-
     Args:
         data (str): string to clean
-
```
Comment (ElieKadoche) [inline clean_dataset.py:19]: Why do you remove these lines? There are part of the docstring.

File: clean_dataset.py [hunk 4]
```diff
@@ -64,6 +70,7 @@ def cleaner(data):
 
     x_train_clean = np.array(x_train_clean)
     x_test_clean = np.array(x_test_clean)
+
```
Comment (ElieKadoche) [inline clean_dataset.py]: Why do you add an empty line here?

File: clean_dataset.py [hunk 5]
```diff
@@ -64,6 +70,7 @@ def cleaner(data):
 
     x_train_clean = np.array(x_train_clean)
     x_test_clean = np.array(x_test_clean)
+    
 
     np.save(utils.X_TRAIN_PATH, x_train_clean)
-    np.save(utils.X_TEST_PATH, x_test_clean)
+    np.save(utils.X_TEST_PATH, x_test_clean)
```
Comment (ElieKadoche) [inline clean_dataset.py]: Typo error.

File: create_datasets.py [hunk 6]
```diff
@@ -35,4 +35,4 @@
     np.save(utils.X_TEST_PATH, x_test_data)
     np.save(utils.Y_TEST_PATH, y_test_data)
 
-    print("Done!")
+    print("Done!")
```
Comment (ElieKadoche) [inline create_datasets.py]: Typo error.

File: download_dataset.py [hunk 7]
```diff
@@ -24,3 +24,4 @@
     z = zipfile.ZipFile(io.BytesIO(r.content))
     z.extractall(utils.DATASET_FOLDER)
     print("Done!")
+
```
Comment (ElieKadoche) [inline download_dataset.py]: Typo error, no empty line at the end.

File: model.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Thu May  7 19:32:15 2020
+
+@author: minh
```
Comment (ElieKadoche) [inline model.py]: Please remove these, this is a common project.

File: model.py [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+"""
+
+"""First model for the sentiment analysis problem.
+This file can be used as a model for all models.
```
Comment (ElieKadoche) [inline model.py]: This is true for the template, wrong the this model.

File: model.py [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        """Init function.
+
+        vocab_size - vocabulary size
+        embedding_dim - size of the dense word vectors
+        hidden_dim - size of the hidden states
+        output_dim - number of classes
+        n_layers - number of multi-layer RNN
+        bidirectional - boolean - use both directions of LSTM
+        dropout - dropout probability
+        pad_idx -  string representing the pad token
```
Comment (ElieKadoche) [inline model.py]: Please respect Google docstring types. [Here](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) an example.

File: model.py [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+        
+        # Fully-connected layer
+        self.predictor = nn.Linear(hidden_dim*2, output_dim)
+
+        # Initialize dropout layer for regularization
+        self.dropout = nn.Dropout(dropout)
+      
+    def forward(self, x, x_lengths):
+        """Forward function.
+        Args:
```
Comment (ElieKadoche) [inline model.py]: Empty line between description and Args.

File: model.py [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+            x (tensor): tensor of strings, shape (N,) with N the batch size
+            s (tensor): tensor of scalars, shape (N,) with N the batch size
+        """
+        embedded = self.dropout(self.embedding(x))    
+
+        packed_embedded = nn.utils.rnn.pack_padded_sequence(embedded, x_lengths) #pack the embeddings
```
Comment (ElieKadoche) [inline model.py]: For the comment, you need a space after the "#" and you need to begin your comment by a capital letter. I will not mark all of them in the code, but same commentary.

File: model.py [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+
+        hidden = self.dropout(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim = 1)) #final layer forward and backward hidden states
+
+        return self.predictor(hidden)
+
+
+def create(model):
+    """Create a fresh new model.
+    This function does not train or test the model. It just creates it and
+    save it in models/model_name/model_name.pt.
```
Comment (ElieKadoche) [inline model.py]: Empty line + args.

File: model.py [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+    """
+    logging.info("launching create function")
+
+    #model = Model().to(device)
+    #model = LSTM()
```
Comment (ElieKadoche) [inline model.py]: Do not keep commented lines in your code. It is either relevant or it is not.

File: model.py [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+    torch.save(model.state_dict(), MODEL_PATH)
+
+    logging.info("model created and saved in {}".format(MODEL_PATH))
+    
+def batch_accuracy(predictions, label):
+
```
Comment (ElieKadoche) [inline model.py]: No empty line here...

File: model.py [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+    #model = LSTM()
+    torch.save(model.state_dict(), MODEL_PATH)
+
+    logging.info("model created and saved in {}".format(MODEL_PATH))
+    
+def batch_accuracy(predictions, label):
+    
+    """Compute accuracy per batch.
+    predictions - float
+    label - 0 or 1
```
Comment (ElieKadoche) [inline model.py]: Google docstring...

File: model.py [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+
+    return accuracy
+
+
+def train(model, iterator, optimizer, criterion):
+    """Train an existing model.
+    This function trains the model models/model_name/model_name.pt. It does not
+    use the testing dataset, but only the training one. The model is updated
+    after each epoch.
+    All operations on the training dataset are allowed.
```
Comment (ElieKadoche) [inline model.py]: Again.

File: model.py [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+    """
+    logging.info("launching train function")
+    
+
+
```
Comment (ElieKadoche) [inline model.py]: Why so many empty lines?

File: model.py [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+                format="csv",
+                fields=fields,
+                skip_header=False)
+        os.remove('test_set.csv')
+
+    return TEXT, LABEL, dataset
+    
+
+
+
```
Comment (ElieKadoche) [inline model.py]: Why so many empty lines?

File: preload.py [hunk 20]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Fri May  8 21:09:58 2020
+
+@author: minh
+"""
```
Comment (ElieKadoche) [inline preload.py]: What is the purpose of this file?

**Review by ElieKadoche — CHANGES_REQUESTED** [review]
Your code lacks of consistency. You need to respect Google docstring type and be carefull with typo error. You have to explain more your code in the docstring.

The modification on the clean_dataset requires another pull request I think, I found another way to do it.

Do you have any results with your model?

File: preload.py (hunk 20 above)
Comment (ElieKadoche) [reply preload.py]: Ma fault, you said it in your commentary. We should not create another file but do it directly in the clean_dataset.py script.

**Comment by fmdinh** [comment]
Thanks for your remark !

For information, I used an LSTM model. I get an accuracy of 70% for training but only 54% for tests. I think that is also the structure of the LSTM. In fact, LSTM takes words from the transition sequence, from which I removed the punctuation (if I keep the punctuation, the result will be really worse). It may create a little overfit when the sentences get long due to lack of punctuation.

I think that with the CNN model or your other models will have better results!

Also, I think we should start reporting. May I create a link on the overleaf to share for three of us?

Thank you,

**Comment by ElieKadoche** [comment]
As we discussed, I made some modifications. There are still modifications to do on your script, and important ones... I try to do it now.

You were absolutely right, there were empty samples. Good call.

**Comment by ElieKadoche** [comment]
I made some modifications, but your LSTM model script contains too many architecture issues. Please try to correct them... In the `if __name__ == "__main__"` you should not put your code, only arguments and variables shared but the 3 functions. All your code need to be in the dedicated functions. You can of course create as many functions you like, and then use them in the main ones.

Also try to use better class than `torchtext.data.TabularDataset`. Create a csv is totaly useless since you can directly call `torchtext.data.Dataset`, etc.

**Review by ElieKadoche — APPROVED** [review]
